### ISSUE TITLE
flynn-host fails to fully restart after node reboot

### ISSUE THREAD

**Issue by sleighter** [flynn/flynn#3451]
To test automatic healing, I rebooted one node of a three-node cluster. When the node came back online, flynn-host started as expected: `sudo status flynn-host` shows 'running'. However, the local webserver on :1111 is only returning empty responses. This breaks most of the flynn-host commands available on that node and causes the cluster's discoverd process to log a continuous stream of connection failures.

On the restarted node:
```
ubuntu@ip-10-210-18-37:~$ flynn-host list
19:09:06.171373 host.go:154: Error sending HTTP request, errors: Get http://127.0.0.1:1111/services/flynn-host/instances: EOF
```

In the discoverd process logs, this message repeats several times per second:
```
[2016-09-30T19:12:30.114734548Z] 2016/09/30 19:12:30 [DEBUG] raft: Failed to contact 10.210.18.37:1111 in 2h45m59.891473048s
```

Just to be certain, I have confirmed that the restarted node is reachable by both other nodes, but attempts to `curl` the restarted node on port 1111 result in the same empty response.
```
ubuntu@ip-10-210-18-54:~$ curl http://10.210.18.37:1111/services/flynn-host/instances
curl: (52) Empty reply from server
```

**Comment by bbaptist** [comment]
Looks pretty similar to my testing #3111

**Comment by sleighter** [comment]
I was able to recover the node through this process:
 1. SSH to problematic node
 2.`kill -2` the discoverd process
 3. Start discoverd 
    * `cd /var/lib/docker/aufs/mnt/tmp-ip102101837-6ab43c7d-129f-43c3-9470-81e5a2c56755`
    * `bin/discoverd -data-dir=data -host=10.210.18.37 -peers=10.210.18.163:1111,10.210.18.37:1111,10.210.18.54:1111, -addr=:1111 -notify=http://10.210.18.37:1113/host/discoverd -wait-net-dns=true`
 
The cluster immediately picked up the node, put it back into service, and elected a new leader. 

I noticed that when the flynn-host daemon was starting, two jobs were being resurrected from a previous state:
```
2016/09/30 21:26:39 resurrecting ip102101837-f4f4fc65-8b48-4a11-9e96-a4888f73d576 as ip102101837-f0545c02-31ed-45f1-880d-eec6c283146d
2016/09/30 21:26:39 resurrecting ip102101837-5df3e49e-9eab-4cc7-949b-7f5cf3b30c4c as ip102101837-6ab43c7d-129f-43c3-9470-81e5a2c56755
```
One of those was discoverd (the other was containerinit), so I assumed there was something wrong with the state it was restoring. This makes some sense, as the job state was probably corrupted by the system halt. Starting discoverd outside the container, without that state, resolved the issue. 

It would be great to just detect this condition and restart the job cleanly. But in this case, the restored (and failing) discoverd job didn't provide any indication of a failure. The process was running, it was ostensibly bound to right port, but requests never reached the process and clients always received empty responses. 

As mentioned in #3111 , discoverd did not stop cleanly during the reboot, which probably led to this bad state being resurrected. This was logged at the time of the reboot:
```
t=2016-09-30T16:26:29+0000 lvl=info msg="stopping all jobs except discoverd" app=host pid=2270 fn=runDaemon host.id=ip102101837
t=2016-09-30T16:26:32+0000 lvl=info msg="stopping discoverd" app=host pid=2270 fn=runDaemon host.id=ip102101837
t=2016-09-30T16:26:32+0000 lvl=eror msg="error stopping discoverd" app=host pid=2270 fn=runDaemon host.id=ip102101837 err=nil
```

Since discoverd seems to have problems stopping cleanly, could it be forced to always start cleanly and not be resurrected by the daemon?

**Comment by titanous** [comment]
@sleighter Can you post the discoverd log? `sudo flynn-host collect-debug-info` should grab it.

**Comment by sleighter** [comment]
@titanous https://gist.github.com/sleighter/3f7d1993b91d8e731f71f22a3f6f62c4 is the log corresponding to the resurrected job referenced above.

**Comment by titanous** [comment]
@sleighter Sorry, I should have been more specific, I'm also looking for the logs from the other instances of discoverd, including the one that was shut down.

**Comment by sleighter** [comment]
@titanous `collect-debug-info` did not contain logs for the job killed at 16:26. I had grabbed logs earlier, but no job info was included, only `flynn-host.log` and `0-debug-output.log`

These logs were collected shortly after the reboot:
https://s3-us-west-1.amazonaws.com/uat-flynn-blobstore/flynn-host-debug.tar

These logs were collected after many attempts to bring the node back up, so there are lots of crashed process logs:
https://s3-us-west-1.amazonaws.com/uat-flynn-blobstore/flynn-host-debug-2.tar.gz

**Comment by titanous** [comment]
Thanks! We were unable to reproduce this last time it was reported, but clearly it is not as rare as we thought. We will attempt to replicate the issue and sort it out.

**Comment by sleighter** [comment]
@titanous FWIW I seem to be able to replicate this at will. Any time I perform a Reboot of this EC2 instance node, it will fail to re-join. The behavior matches what I saw on Friday. I have found that if I ssh to the node and perform these steps, it will come back online:
 1. `sudo service flynn-host stop`
 2. `sudo flynn-host destroy-volumes`
 3. `sudo service flynn-host start`

**Comment by lmars** [comment]
@sleighter after testing multiple failure scenarios, we don't seem to be able to reproduce this.

What would be useful is if you could trigger this scenario (i.e. where discoverd is returning empty responses) and then send the problematic discoverd process a `SIGQUIT` signal which will cause it to print a stack trace and then exit:

```
$ sudo killall -QUIT discoverd
```

Once you have done that, please upload the contents of the following file from each host in the cluster:

```
/var/log/flynn/b81640a2-6af5-40cb-aa87-b79bb27399c4.log
```

**Comment by sleighter** [comment]
@lmars : After the `killall`, a new discoverd process came up and rejoined the cluster successfully. 

Nonetheless, this is the log from the killed process: https://s3-us-west-1.amazonaws.com/uat-flynn-blobstore/b81640a2-6af5-40cb-aa87-b79bb27399c4.log

**Comment by lmars** [comment]
@sleighter can you also get the logs from the other nodes in the cluster?

**Comment by sleighter** [comment]
@lmars 
Node 1: https://s3-us-west-1.amazonaws.com/uat-flynn-blobstore/8a1b0808-3137-4a4c-b42e-b7e26b7efe8a_node1.log
Node 2: https://s3-us-west-1.amazonaws.com/uat-flynn-blobstore/8a1b0808-3137-4a4c-b42e-b7e26b7efe8a_node2.log

**Comment by lmars** [comment]
@sleighter ok thanks, I think I have a handle on what is going on. I will try and reproduce then work on a patch, I'll keep you posted.

**Comment by lmars** [comment]
@sleighter I have proposed a fix in #3467.

**Comment by SalvoLunar** [comment]
Hi @lmars @titanous 

I'm testing my cluster (3 nodes) and it tried to force shutdown one of my nodes to check if the cluster would still stay up.

The 2 nodes are still working but I'm now trying to figure out why on the node I've restarted it's not possible to restart the flynn-host service.

This is the error I'm getting, do you have any idea of why the restart would fail?

Thanks,
Salvo

```
* flynn-host.service - Flynn host daemon
   Loaded: loaded (/lib/systemd/system/flynn-host.service; enabled; vendor preset: enabled)
   Active: inactive (dead) (Result: exit-code) since Tue 2018-06-19 15:18:26 UTC; 1s ago
     Docs: https://flynn.io/docs
  Process: 10005 ExecStart=/usr/local/bin/flynn-host daemon (code=exited, status=2)
 Main PID: 10005 (code=exited, status=2)

Jun 19 15:18:26 fsphost1.lunarct.com systemd[1]: flynn-host.service: Unit entered failed state.
Jun 19 15:18:26 fsphost1.lunarct.com systemd[1]: flynn-host.service: Failed with result 'exit-code'.
Jun 19 15:18:26 fsphost1.lunarct.com systemd[1]: flynn-host.service: Service hold-off time over, scheduling restart.
Jun 19 15:18:26 fsphost1.lunarct.com systemd[1]: Stopped Flynn host daemon.
Jun 19 15:18:26 fsphost1.lunarct.com systemd[1]: flynn-host.service: Start request repeated too quickly.
Jun 19 15:18:26 fsphost1.lunarct.com systemd[1]: Failed to start Flynn host daemon.
```
